lib/libzfs/libzfs_pool.c

```diff
@@ -2316,49 +2316,86 @@ zpool_trim(zpool_handle_t *zhp, pool_trim_func_t cmd_type, nvlist_t *vds,
 {
-	zfs_cmd_t zc = {"\0"};
 	char msg[1024];
 	int err;
 	libzfs_handle_t *hdl = zhp->zpool_hdl;
 
-	(void) strlcpy(zc.zc_name, zhp->zpool_name, sizeof (zc.zc_name));
-	zc.zc_cookie = func;
-	zc.zc_flags = cmd;
-
-	if (zfs_ioctl(hdl, ZFS_IOC_POOL_SCAN, &zc) == 0)
-		return (0);
+	nvlist_t *args = fnvlist_alloc();
+	fnvlist_add_uint64(args, "scan_type", (uint64_t)func);
+	fnvlist_add_uint64(args, "scan_command", (uint64_t)cmd);
 
-	err = errno;
+	err = lzc_scrub(ZFS_IOC_POOL_SCRUB, zhp->zpool_name, args, NULL);
+	fnvlist_free(args);
 
-	/* ECANCELED on a scrub means we resumed a paused scrub */
-	if (err == ECANCELED && func == POOL_SCAN_SCRUB &&
-	    cmd == POOL_SCRUB_NORMAL)
+	if (err == 0) {
 		return (0);
+	}
 
-	if (err == ENOENT && func != POOL_SCAN_NONE && cmd == POOL_SCRUB_NORMAL)
+	/*
+	 * ECANCELED on a scrub means one of the following.
+	 * 1. we resumed a paused scrub.
+	 * 2. we resumed a error paused scrub.
+	 * 3. Error scrub is not run because of no error log.
+	 */
+	if (err == ECANCELED && (func == POOL_SCAN_SCRUB ||
+	    func == POOL_ERRORSCRUB) && cmd == POOL_SCRUB_NORMAL)
 		return (0);
+	/*
+	 * Following cases have been handelled here.
+	 * 1. Paused a scrub/error scrub if there is no in progross.
+	 */
+	if (err == ENOENT && func != POOL_SCAN_NONE && (cmd ==
+	    POOL_SCRUB_NORMAL || cmd != POOL_ERRORSCRUB_STOP)) {
+		return (0);
+	}
 
 	if (func == POOL_SCAN_SCRUB) {
 		if (cmd == POOL_SCRUB_PAUSE) {
 			(void) snprintf(msg, sizeof (msg), dgettext(TEXT_DOMAIN,
-			    "cannot pause scrubbing %s"), zc.zc_name);
+			    "cannot pause scrubbing %s"), zhp->zpool_name);
 		} else {
-			assert(cmd == POOL_SCRUB_NORMAL);
+			ASSERT3U(cmd, ==, POOL_SCRUB_NORMAL);
+			(void) snprintf(msg, sizeof (msg), dgettext(TEXT_DOMAIN,
+			    "cannot scrub %s"), zhp->zpool_name);
+		}
+	} else if (func == POOL_ERRORSCRUB) {
+		if (cmd == POOL_SCRUB_PAUSE) {
 			(void) snprintf(msg, sizeof (msg), dgettext(TEXT_DOMAIN,
-			    "cannot scrub %s"), zc.zc_name);
+			    "cannot pause error scrubbing %s"),
+			    zhp->zpool_name);
+		} else if (cmd == POOL_SCRUB_NORMAL) {
+			(void) snprintf(msg, sizeof (msg), dgettext(TEXT_DOMAIN,
+			    "cannot error scrub %s"), zhp->zpool_name);
+		} else {
+			ASSERT3U(cmd, ==, POOL_ERRORSCRUB_STOP);
+			(void) snprintf(msg, sizeof (msg),
+			    dgettext(TEXT_DOMAIN, "cannot cancel error "
+			    "scrubbing %s"), zhp->zpool_name);
 		}
 	} else if (func == POOL_SCAN_RESILVER) {
-		assert(cmd == POOL_SCRUB_NORMAL);
+		ASSERT3U(cmd, ==, POOL_SCRUB_NORMAL);
 		(void) snprintf(msg, sizeof (msg), dgettext(TEXT_DOMAIN,
-		    "cannot restart resilver on %s"), zc.zc_name);
+		    "cannot restart resilver on %s"), zhp->zpool_name);
 	} else if (func == POOL_SCAN_NONE) {
 		(void) snprintf(msg, sizeof (msg),
 		    dgettext(TEXT_DOMAIN, "cannot cancel scrubbing %s"),
-		    zc.zc_name);
+		    zhp->zpool_name);
 	} else {
 		assert(!"unexpected result");
```

Comment:
I'd get rid of this last `else` block, and move the assert() to before the `if (func == POOL_SCAN_SCRUB)` block, with this check:
`assert(func > POOL_SCAN_NONE  && func < POOL_SCAN_FUNCS);`